### PR TITLE
Migrate to new Google Cloud impersonation API

### LINKED ISSUE #83: Broken Service Account impersonation
#### Overview of the Issue

When using [Google Cloud IAM impersonation](https://cloud.google.com/iam/docs/impersonating-service-accounts) mechanism I'm getting an error stating that my user doesn't have sufficient privileges.
I successfully used this mechanism with Terraform hoping it would be the same with Packer.

#### Reproduction Steps

First of all, in order to use the impersonation mechanism, it is necessary to configure the appropriate service accounts as described [here](https://cloud.google.com/iam/docs/impersonating-service-accounts):
1. Create the Service Account that owns the permissions to interact with the GCE instances:
    1. To make the tests easier, assign the role `roles/editor` to it;
    2. Service account identifier: `service_account_builder@your_project_id.iam.gserviceaccount.com`;
2. Create the Service Account that will be used to impersonate the service account above:
    1. Assign no permission to it;
    2. Service account identifier: `service_account_doorman@your_project_id.iam.gserviceaccount.com`;
3. Allow `service_account_doorman@your_project_id.iam.gserviceaccount.com` to impersonate `service_account_builder@your_project_id.iam.gserviceaccount.com` by granting him the roles below:
        1. `Service Account Token Creator`;
        2. `Service Account User`.
4. Finally, generate a JSON key for the service account `service_account_doorman@your_project_id.iam.gserviceaccount.com`: `impersonation_json_key.json`.

Finally, use the JSON key to run the build.

### Plugin and Packer version

```bash
$ packer --version
1.7.10
```

### Simplified Packer Buildfile

```
packer {
  required_plugins {
    googlecompute = {
      version = ">= 1.0.10"
      source  = "github.com/hashicorp/googlecompute"
    }
  }
}

source "googlecompute" "vm_image" {
  project_id                  = "your_project_id"
  source_image_family         = "centos-7"
  zone                        = "europe-west1-b"
  machine_type                = "n1-standard-8"
  impersonate_service_account = "service_account_doorman@your_project_id.iam.gserviceaccount.com"
  image_description           = "VM image"
  instance_name               = "tpl-vm-image-{{uuid}}"
  disk_size                   = 50
  disk_type                   = "pd-ssd"
  metadata = {
    enable-oslogin : "true"
  }
  use_os_login      = true
  ssh_username      = "root"
  image_name        = "tpl-vm-image"
}

build {
  sources = ["sources.googlecompute.vm_image"]

  provisioner "shell" {
    inline = ["echo foo"]
  }
}
```
And finally, run the build:

```bash
GOOGLE_APPLICATION_CREDENTIALS="impersonation_json_key.json" packer build template.pkr.hcl
```

### Operating system and Environment details

```bash
$ uname -a
Linux 3.10.0-1160.53.1.el7.x86_64 #1 SMP Fri Jan 14 13:59:45 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/*-release
CentOS Linux release 8.4.2105
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
CentOS Linux release 8.4.2105
CentOS Linux release 8.4.2105
```

### Log Fragments and crash.log files

Full logs available [here](https://gist.github.com/fgateuil/32e977f5fc902c43fb3eded1e6f0f163).

### PR DESCRIPTION
**Pull request by fgateuil**
I updated the code using the non-deprecated way of using Google Cloud IAM impersonation.

Despite my efforts, I've been unable to submit any relevant unit tests since this part of the code is tied to Google remote API (POST requests are performed).
I tried to invoke/create mocking objects without success.

As a result, I think that this should be tested with integration tests but it looks like nothing similar exists yet (it would probably be quite expensive to have a dedicated GCP project).

Nevertheless, I would be more than happy if someone could help me on that.
Hope my contribution will be helpful.

Closes #83

### CODE REVIEW

**Review by azr — APPROVED** [review]
Hey, @fgateuil thanks for making this happen, this LGTM, we have acceptance tests that run periodically and that should cover this. I'll add a task in our backlog.

File: builder/googlecompute/driver_gce.go [hunk 1]
```diff
@@ -92,7 +93,18 @@ func NewClientOptionGoogle(account *ServiceAccount, vaultOauth string, impersona
 		opts = append(opts, option.WithTokenSource(ts))
 
 	} else if impersonatesa != "" {
-		opts = append(opts, option.ImpersonateCredentials(impersonatesa))
+		log.Printf("[INFO] Using Google Cloud impersonation mechanism")
+		ts, err := impersonate.CredentialsTokenSource(context.Background(), impersonate.CredentialsConfig{
+			TargetPrincipal: impersonatesa,
+			Scopes: []string{
+				"https://www.googleapis.com/auth/cloud-platform",
+				"https://www.googleapis.com/auth/userinfo.email",
```
Comment (azr) [inline builder/googlecompute/driver_gce.go:101]: Can you tell me more about these scopes, should these be parameters ? Are they the sort of minimal requirement ?

Comment (fgateuil) [reply builder/googlecompute/driver_gce.go:101]: This is clearly an unwanted error on my part. I'm gonna fix that immediatly as it should not be hard-coded this way.
Thanks for the warning  ;)

Comment (fgateuil) [reply builder/googlecompute/driver_gce.go]: > Can you tell me more about these scopes, should these be parameters ? Are they the sort of minimal requirement ?

@azr change is now done. Please tell me if you think additional work is needed. Thanks.

File: post-processor/googlecompute-import/post-processor.go [hunk 2]
```diff
@@ -36,6 +36,17 @@ type Config struct {
...
 	// This allows service account impersonation as per the [docs](https://cloud.google.com/iam/docs/impersonating-service-accounts).
 	ImpersonateServiceAccount string `mapstructure:"impersonate_service_account" required:"false"`
+	// The service account scopes for launched
+	// instance. Defaults to:
+	//
+	// ```json
+	// [
+	//   "https://www.googleapis.com/auth/userinfo.email",
+	//   "https://www.googleapis.com/auth/compute",
+	//   "https://www.googleapis.com/auth/devstorage.full_control"
```
Comment (azr) [inline post-processor/googlecompute-import/post-processor.go:46]: Hello there thanks for updating this and no worries . :)

Another question here, where is this default happening here ? In the plugin codebase or over on GCP side ?

- If it is the first case, then I haven't found the code doing that.
- If it is the latter, then I think that we should point at the remote docs here.

Comment (fgateuil) [reply post-processor/googlecompute-import/post-processor.go:46]: Ok, I got it now. I finally committed sin out of ignorance.
I was totally unaware of the other options that the plugin offers (particularly the post-process `import` and `export`). My comments in the code are indeed wrong.
I will make additional tests and check that these use-cases are well covered, and finally will update the code/doc accordingly.

**Comment by fgateuil** [comment]
If everything is now ok for you @azr, feel free to merge the PR  ;)
Thanks.

**Review by nywilken — APPROVED** [review]
Updated changes looks good to me. Piggy backing off of @azr initial review. Thanks for the contribution.

**Pull request merged**